share/availability/light/availability.go

```diff
@@ -66,6 +64,17 @@ func (la *ShareAvailability) SharesAvailable(ctx context.Context, header *header
 		return nil
 	}
 
+	// Lock to prevent multiple sampling sessions for the same header height
+	lk, loaded := la.inProgress.LoadOrStore(header.Height(), &sync.Mutex{})
+	lk.(*sync.Mutex).Lock()
+	defer func() {
+		lk.(*sync.Mutex).Unlock()
+		// Delete the lock if it was created in this call
+		if !loaded {
+			la.inProgress.Delete(header.Height())
+		}
+	}()
+
```

Comment:
If we ever make SharesAvailable blocking (we will), it would be great to have this respecting context. It shouldn't be hard to change Mutex to a channel. This will partially resemble `heightSub` logic and can be directly copied from there.

Reply:
Good idea. Reworked to use channels. After visiting `heightSub` it seems unnecessary complicated, so I've similar, but simpler solution. Perhaps it can be used in heightSub too?